core/src/test/scala/integration/kafka/api/EndToEndClusterIdTest.scala

```diff
@@ -0,0 +1,204 @@
+/**
+  * Licensed to the Apache Software Foundation (ASF) under one or more
+  * contributor license agreements.  See the NOTICE file distributed with
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  * http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.api
+
+import java.util.concurrent.atomic.AtomicReference
+import java.util.{ArrayList, Properties}
+
+import kafka.common.TopicAndPartition
+import kafka.integration.KafkaServerTestHarness
+import kafka.server._
+import kafka.utils._
+import org.apache.kafka.clients.consumer._
+import org.apache.kafka.clients.producer.{KafkaProducer, ProducerConfig, ProducerRecord}
+import org.apache.kafka.common.{ClusterResource, ClusterResourceListener, TopicPartition}
+import org.apache.kafka.test.{TestUtils => _, _}
+import org.junit.Assert._
+import org.junit.{Before, Test}
+
+import scala.collection.JavaConverters._
+
+/** The test cases here verify the following:
+  * 1. The ProducerInterceptor receives the Cluster metadata after first send.
```

Comment:
Ok, now that you mention it, I agree that it will be weird to the users who don't know how the client work internally. 

Do you think I should only call the listener only if cluster id is present ? I can add that condition in the Metadata class. (I can't think of any reason the users will want to get a notification for the bootstrap cluster).